staging/src/k8s.io/apimachinery/pkg/runtime/converter.go

```diff
@@ -358,39 +378,86 @@ func pointerFromUnstructured(sv, dv reflect.Value) error {
 	dv.Set(reflect.New(dt.Elem()))
 	switch st.Kind() {
 	case reflect.Ptr, reflect.Interface:
-		return fromUnstructured(sv.Elem(), dv.Elem())
+		return c.fromUnstructured(sv.Elem(), dv.Elem())
 	default:
-		return fromUnstructured(sv, dv.Elem())
+		return c.fromUnstructured(sv, dv.Elem())
 	}
 }
 
-func structFromUnstructured(sv, dv reflect.Value) error {
-	st, dt := sv.Type(), dv.Type()
-	if st.Kind() != reflect.Map {
-		return fmt.Errorf("cannot restore struct from: %v", st.Kind())
-	}
-
+// flattenedFields takes a value and returns all inlined fields as top
+// level fields in order to determine which fields are invalid.
+func flattenedFields(dv reflect.Value) map[reflect.Value]reflect.Value {
```

Comment:
Okay, I believe I have implemented the refactor as you have described @liggitt 

We are now threading the an arg to accumulate the unknown fields via `fromUnstructuredOptions` that handles collecting the matched keys.

@apelisse I do think this is the better approach now because otherwise it is hard to collect the full path to the unknown fields, and by passing the sets of errors/matched fields via the options I don't think we are allocating as many slices/maps as we were previously.